_posts/2021-12-29-readable-tests-by-example.md

```diff
@@ -0,0 +1,484 @@
+  void assertThatClientDidNotPaidForDelivery(OrderPaidEventBuilder anEventBuilder = anOrderPaidEvent().anOrderPaidEventWithFreeDelivery()) {
+    pollingConditions.eventually {
+      Mockito.verify(domainEventPublisher, times(1))
+        .publish(anEventBuilder.build())
+    }
+  }
+  // some code omitted
+}
+```
+
+The question is why we should make so much effort in creating own solutions, and not use ready-made solutions directly from the framework? Here are the arguments for:
+* reusability - we can use once written ability in many places,
+* extensibility - in the case of changing the library, which, for example, is used to mock other services, it is enough to make changes in one place,
+* enriching the test with the language specific to our domain,
+* we are not limited by the capabilities of a given framework, e.g. Spock doesn’t allow you to mock final Java classes, then we have to use an additional lib like [spock-mockable](https://github.com/CamilYed/readable-tests-by-example/blob/master/src/test/groovy/tech/allegro/blog/vinyl/shop/delivery/domain/DeliveryCostPolicySpec.groovy#L13).
+
+### Tailor-made assertions
+
+The last concept that I want to discuss is dedicated assertion classes. Assertion class is nothing more than a simple class exposing methods which allow checking the input object appropriately.
+
+In some scenarios, we would like in fact to verify the data that, for example, was eventually saved in the database.
+Some of our objects can be so complex that it would be inconvenient to check them directly in the test, referencing the nested objects or iterating the collections.
+
+```groovy
+  def "should change the item quantity for unpaid order"() {
+    given:
+        thereIs(anUnpaidOrder()
+                .withId(ORDER_ID)
+                .withClientId(CLIENT_ID)
+                  .withItems(
+                    anItem()
+                      .withProductId(CZESLAW_NIEMEN_ALBUM_ID)
+                      .withUnitPrice(euro(35.00))
+                      .withQuantity(10),
+                    anItem()
+                      .withProductId(BOHEMIAN_RHAPSODY_ALBUM_ID)
+                      .withUnitPrice(euro(55.00))
+                      .withQuantity(1)
+                  )
+        )
+
+    when:
+       changeItemQuantity(anItemQuantityChange()
+                            .withOrderId(ORDER_ID)
+                            .withProductId(CZESLAW_NIEMEN_ALBUM_ID)
+                            .withQuantityChange(20)
+       )
+
+    then:
+        assertThatThereIsOrderWithId(ORDER_ID)
+          .hasClientId(CLIENT_ID)
+          .hasItemWithIdThat(CZESLAW_NIEMEN_ALBUM_ID)
+              .hasUnitPrice(euro(35.00))
+              .hasQuantity(20)
+          .and()
+          .hasItemWithIdThat(BOHEMIAN_RHAPSODY_ALBUM_ID)
+              .hasUnitPrice(euro(55.00))
+              .hasQuantity(1)
+  }
+```
+Apart from this, such an assertion can also be used in other places than just one test class.
+
```

Comment:
It would be nice to mention that these techniques are easily adaptable by other languages, for example: Kotlin.